Initial push of metadata skeleton

DalerHakimov opened the pull request:
First draft developed version of generating metadata package for end user. It generates 3 json files under output directory by reading json files from capture, permissions, visualizations.

DalerHakimov committed:
Initial push of metadata skeleton

DalerHakimov committed:
rm unused

tokland commented on .DS_Store:
We typically don't upload this kind of OS-specific files. 

Add to .gitignore (file that should be pushed to the repo). Get as reference some .gitignore from another Typescript EST project.

tokland commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+## Prerequisites
+
+- Node.js (>=14.x)
```
A .nvmrc file should be added to the project so DEvs can use the specific version easily (running `nvm use`)

tokland commented on README.md:
```diff
@@ -0,0 +1,42 @@
+## Prerequisites
+
+- Node.js (>=14.x)
+- npm (>=6.x) or yarn (>=1.x)
```
As we have a yarn.lock, this is a yarn project, let's remove all references to npm, no need to support both.

tokland commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+1. Clone the repository:
+
+    git clone https://github.com/yourusername/metadata-skeleton-offline-2.git
```
Add sh block

tokland commented on README.md:
````diff
@@ -0,0 +1,42 @@
+    or
+
+    ```sh
+    yarn install
````
Add "nvm use" as first command, as we'll have now a .nvmrc file

tokland commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+To start the project, run the following command:
+
+yarn start
```
sh block

tokland commented on package.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "metadata-skeleton-offline-2",
```
why the 2?

tokland commented on src/domain/entities/Category.ts:
```diff
@@ -0,0 +1,8 @@
+    dimensionType: string;
+    categoryOptions: { id: string }[];
+    dataDimension: boolean;
+}
```
That red icon mean "No newline at end of file". All text files should have it. Configure your editor so it's automatically added on save.

[Same for all other files in the project]

tokland commented on src/domain/entities/Category.ts:
```diff
@@ -0,0 +1,8 @@
+export interface Category extends BaseMetadata {
+    dataDimensionType: string;
+    dimensionType: string;
+    categoryOptions: { id: string }[];
```
`{ id: string }` is very common, we typically define a type alias called "Ref" and use everywhere it's needed.

tokland commented on src/domain/entities/CategoryCombos.ts:
```diff
@@ -0,0 +1,13 @@
+export interface CategoryCombos extends BaseMetadata {
+    categories: Array<{
+        id: string;
+    }>;
```
As we've seen: `categories: Ref[]` (refactor all references in the project)

tokland commented on src/domain/entities/CategoryCombos.ts:
```diff
@@ -0,0 +1,13 @@
+    categories: Array<{
+        id: string;
+    }>;
+    dataDimensionType: string;
```
Can we be more specific? `"DISAGGREGATION" | "ATTRIBUTE"`

tokland commented on src/domain/entities/CategoryOption.ts:
```diff
@@ -0,0 +1,12 @@
+export interface CategoryOption extends BaseMetadata {
+    startDate?: string;
+    endDate?: string;
+    organisationUnits: any[];
```
As a rule of thumb, we don't ever use "any" as type. If we are not going to interact with that field, we remove it or put it as "unknown".

We are adding a lot of fields in metadata that we don't interact with. Typically we don't do that, but ok, now that's done, it won't hurt.

tokland commented on src/domain/repository/FileRepository.ts:
```diff
@@ -0,0 +1,42 @@
+        this.folderName = path.basename(capturePath);
+    }
+
+    async validateFiles(): Promise<string | null> {
```
The repositories in the domain only define the `interface` the methods that will be available. The implementation is then done in the data layer (src/data). For example, a DataSetController will have a DataSetD2Controller, which implements it for DHIS2.

Check another EST projects for more examples. I.e: https://github.com/EyeSeeTea/dataset-configuration/tree/development-project-configurator

tokland commented on src/domain/repository/FileRepository.ts:
```diff
@@ -0,0 +1,42 @@
+import fs from 'fs/promises';
+import path from 'path';
+
+export class FileRepository {
```
In Clean Arch, we don't define generic repositories like this. This could be a helper class within the data layer, but not a repository. **A repository is associated with a domain entity**. For example, if we have entity DataSet and we want to persist the dataSets, we will create a DataSetRepository. 

You can call this a "service", for example.

tokland commented on src/domain/repository/FileRepository.ts:
```diff
@@ -0,0 +1,42 @@
+        return files.filter(file => 
+            file.endsWith('.json') && 
+            !file.includes('-sorted')
+        );
```
Files are not automatically formatted. What we do:

- Add .prettierrc
- Add task "yarn prettier"
- Configure the editor (vscode?) to apply format on save.

You an check other projects in EST for reference.

tokland commented on src/domain/repository/FileRepository.ts:
```diff
@@ -0,0 +1,42 @@
+        );
+    }
+
+    async readJsonFile(filename: string): Promise<any> {
```
Instead of returning "any", when we read from external sources, we may cast to the expected type:

`async readJsonFile<T>(filename: string): Promise<T> { ... }`

And we use it like this: `readJsonFile<{name: string}>(....)`

On more advanced cases, we check the type with zod or purify-ts/Codec, but I don't think it's needed here.

tokland commented on src/domain/useCases/DataSetProgramCase.ts:
```diff
@@ -0,0 +1,72 @@
+            for (const file of files) {
+                const parsedData = await this.fileRepository.readJsonFile(file);
+                this.combineData(processedData, parsedData);
+            }
```
We use functional/immutable style (as opposed to imperative style). No mutations.

The iterative process you are doing here is a "folding", which in JS it's done with Array.reduce. A refactor proposal:

```
            const processedData = parsedData.reduce((acc, processed) => {
                return {
                    ...acc,
                    dataSets: concatUnique(acc.dataSets, processed.dataSets),
                    // the rest of props to concat
                }
            })
```

Having defined this somewhere in the file:

```
function concatUnique<T extends Ref>(objs1: T[], objs2: T[]): T[] {
    return _(objs1).concat(objs2).uniqBy(obj => obj.id).value();
}
```

tokland commented on src/domain/useCases/DataSetProgramCase.ts:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    private combineData(processedData: ProcessedDataSet, parsedData: any): void {
```
And then we can remove this function, which does not follow good practices: mutable, uses any, Array.isArray and so on.

tokland commented on src/domain/useCases/PermissionCase.ts:
```diff
@@ -0,0 +1,43 @@
+import { ProcessPermissions } from '../entities/ProcessPermissions';
+import { FileRepository } from '../repository/FileRepository';
+
+export class PermissionProcessor {
```
Typically, we align the file with the main class.

tokland commented on src/domain/useCases/PermissionCase.ts:
```diff
@@ -0,0 +1,43 @@
+import { ProcessPermissions } from '../entities/ProcessPermissions';
```
Name of the file: prefix "Case"? "UseCase"? Or Processor?

I am not familiar with the term "Processor" for clean architecture.

tokland commented on src/domain/useCases/PermissionCase.ts:
```diff
@@ -0,0 +1,43 @@
+                });
+            };
+            if (Array.isArray(parsedData.userGroups)) addUniqueItems(processedData.userGroups, parsedData.userGroups);
+            if (Array.isArray(parsedData.users)) addUniqueItems(processedData.users, parsedData.users);
```
same

tokland commented on src/domain/useCases/ProgramUseCase.ts:
```diff
@@ -0,0 +1,56 @@
+        if (Array.isArray(parsedData.programRuleActions)) addUniqueItems(processedData.programRuleActions, parsedData.programRuleActions);
+        if (Array.isArray(parsedData.programRuleVariables)) addUniqueItems(processedData.programRuleVariables, parsedData.programRuleVariables);
+        if (Array.isArray(parsedData.programIndicators)) addUniqueItems(processedData.programIndicators, parsedData.programIndicators);
+    }
```
same

tokland commented on src/domain/useCases/VisualizationCase.ts:
```diff
@@ -0,0 +1,49 @@
+        if (Array.isArray(parsedData.indicatorTypes)) addUniqueItems(processedData.indicatorTypes, parsedData.indicatorTypes);
+        if (Array.isArray(parsedData.dashboards)) addUniqueItems(processedData.dashboards, parsedData.dashboards);
+    }
+}
```
same

tokland commented on src/index.ts:
```diff
@@ -0,0 +1,55 @@
+
+    // Process visualizations
+    const visualizationProcessor = new VisualizationProcessor(visualizationRepository);
+    await visualizationProcessor.process();
```
In Clean Arch, the presentation layer (here, the script) should call a single use case that performs all the tasks, not call N use cases (or processors).

tokland reviewed (commented):
@DalerHakimov  Can you add to the PR description the link to the task?

In-line comments:

Commit:
Add gitignore, .nvrmc, updates on readme install inmutable-map

Commit:
Add Ref, collection, hashmap, changed any into unknown and some prettier fixes

JorgeMMdev committed:
Usecases refactor

JorgeMMdev committed:
UseCases refactor + updated filenames, entities and index

JorgeMMdev committed:
Renamed variables and functions

JorgeMMdev committed:
add concat function, changed any to unknown and changed from appendUnique to concatUnique

JorgeMMdev committed:
Add commands for the scripts

JorgeMMdev committed:
add lint configuration

JorgeMMdev committed:
eslint version fix

JorgeMMdev committed:
Add validation of Json files

JorgeMMdev committed:
Change capturePath to inputPath

JorgeMMdev committed:
remove compare

JorgeMMdev committed:
React App config

JorgeMMdev committed:
Resolve error of the app config

JorgeMMdev committed:
fix app launch

JorgeMMdev committed:
jsondiffpatch lib test

JorgeMMdev committed:
Clean Dhis2 dependencies

JorgeMMdev committed:
remove d2-api dependencies

JorgeMMdev committed:
clean unused components

JorgeMMdev committed:
Basic componentes + upload containers

JorgeMMdev committed:
json containers and improved app structure

JorgeMMdev committed:
improved app structure

deeonwuli committed:
feat: add json difference to comparator box

deeonwuli committed:
feat: update title in index.html and clean up unused code

deeonwuli committed:
feat: start comparator only if jsons are not equal

deeonwuli committed:
feat: refactor JSON comparison logic and implement file download functionality

deeonwuli committed:
feat: enhance JSON comparison functionality with new components and utilities

deeonwuli committed:
feat: add useDiffLineProps hook and refactor DiffLine component for improved state management

deeonwuli committed:
feat: reorganize hooks structure by moving useComparator, useDiffLineProps, and useDownloadJSON to hooks directory

deeonwuli committed:
fix: add error handling for JSON parsing in parseMetadataFromFile function

deeonwuli committed:
refactor: replace hardcoded colors with a colors object in useDiffLineProps hook

deeonwuli committed:
fix: add error handling for JSON parsing in parseJSONFromEnvInput function

deeonwuli committed:
refactor: simplify renderValue function

deeonwuli committed:
chore: add unit tests for renderValue function/fix failing test

deeonwuli committed:
fix: update command argument names for clarity in compare-metadata function

deeonwuli committed:
fix: improve error logging for JSON parsing

deeonwuli committed:
refactor: replace string literals with Choice enum

deeonwuli committed:
fix: replace JSON.parse with structuredClone for deep cloning and improve handling of undefined values in renderValue function

deeonwuli committed:
fix: update width property in FileInputWrapper and ensure handleFileChange dependency is set correctly

deeonwuli committed:
chore: update translation files

deeonwuli committed:
fix: remove unnecessary css property

deeonwuli committed:
feat: implement temporary file handling and refactor JSON fetching logic

deeonwuli committed:
chore: ignore .DS_Store and remove from repo

deeonwuli committed:
fix: update translation strings in MergedPreviewContainer

deeonwuli committed:
fix: update paths in build file to use process.cwd()

deeonwuli committed:
feat: allow different names for output files

derive output filename from the input folder name

deeonwuli committed:
refactor: change JSONContent from class to interface and update related methods

deeonwuli committed:
fix: update path resolution in build command to use path.resolve

deeonwuli committed:
feat: enable comparator only mode in Vite configuration to skip DHIS2 proxy and auth checks

deeonwuli committed:
Refactor JSON Comparator/Merge Component

- Removed DiffLine and DiffViewer components to streamline the codebase.
- Introduced DiffSection component to handle the diffing and merging logic.
- Updated useComparator hook to manage left and right JSON states and merging functionality.
- Implemented useJsonDiffSelector hook for managing JSON differences and selections.
- Enhanced ImportFile component to support optional labels and improved styling.
- Refined LandingPage layout and styling for better user experience.
- Adjusted theme provider integration in App component for consistent theming.

deeonwuli committed:
fix: update logo source in LandingPage component

deeonwuli committed:
refactor: convert function declarations to arrow functions in JSONContent and improve type handling in useJsonDiffSelector
feat: add json utility functions for sorting, formatting, parsing, and cloning JSON data
fix: update logo import in LandingPage and enhance layout structure

deeonwuli committed:
refactor: replace fast-json-patch with custom JSON comparison logic and remove unused ComparatorViewModel component

deeonwuli committed:
chore: install packages, update translation files

deeonwuli committed:
add empty capture directory

deeonwuli committed:
fix: sort array of objects by id first

deeonwuli committed:
fix: setNestedValue function to handle creating nested structures when left path is N/A

deeonwuli committed:
feat: implement restoreOriginalOrder function and update useComparator to maintain original order of merged JSON

deeonwuli committed:
fix: ts terminal error

deeonwuli committed:
docs: update README with metadata comparator usage instructions and formatting improvements

deeonwuli committed:
Add capture folder structure

deeonwuli committed:
fix: update project name and repository links in README and package.json

deeonwuli committed:
test: remove obsolete App test files

deeonwuli committed:
refactor: remove obsolete test scripts from package.json and clean up VisualizationCombineAndRemoveDuplicatesUseCase

deeonwuli committed:
feat: implement metadata comparison command in CLI

deeonwuli committed:
refactor: update dependencies and remove obsolete packages

deeonwuli committed:
fix: improve ESLint configuration and fix dependency handling in useComparator hook

deeonwuli committed:
refactor: convert Share component from class to functional component

deeonwuli committed:
refactor: centralize JSON parsing logic by introducing parseJson utility

deeonwuli committed:
feat: add pre-push hook to automate code formatting, linting, and testing

deeonwuli committed:
feat: add GitHub workflows and pull request template, remove Travis CI configuration

deeonwuli committed:
feat: refactor metadata processing to use Future for async operations

deeonwuli committed:
feat: remove deprecated entity interfaces and replace with MetadataItem type

deeonwuli committed:
feat: update Readme, update compare-metadata command

deeonwuli committed:
fix: update JSON handling to use JSONValue type

deeonwuli committed:
fix: compare and merge json correctly for array of objects

deeonwuli committed:
tests: add tests for JSON utilities

deeonwuli committed:
fix: update compare-metadata command to correct argument order

adrianq committed:
minor update in the README

adrianq committed:
Merge pull request #2 from EyeSeeTea/feature/metadata-comparator-app

[feature]: Metadata package comparator/merger

deeonwuli committed:
chore: update main workflow to skip bundlemon and only include run-tests and codeql-scan
